go/internal/types/extension_types.go

```diff
@@ -209,13 +209,13 @@ func (UUIDType) ExtensionName() string { return "uuid" }
 func (UUIDType) Serialize() string { return "uuid-serialized" }
 
 // Deserialize expects storageType to be FixedSizeBinaryType{ByteWidth: 16} and the data to be
-// "uuid-serialized" in order to correctly create a UuidType for testing deserialize.
+// "uuid-serialized" in order to correctly create a UUIDType for testing deserialize.
 func (UUIDType) Deserialize(storageType arrow.DataType, data string) (arrow.ExtensionType, error) {
 	if string(data) != "uuid-serialized" {
 		return nil, fmt.Errorf("type identifier did not match: '%s'", string(data))
 	}
-	if !arrow.TypeEqual(storageType, &arrow.FixedSizeBinaryType{ByteWidth: 16}) {
-		return nil, fmt.Errorf("invalid storage type for UuidType: %s", storageType.Name())
+	if !arrow.StorageTypeEqual(storageType, &arrow.FixedSizeBinaryType{ByteWidth: 16}) {
+		return nil, fmt.Errorf("invalid storage type for UUIDType: %s", storageType.Name())
```

Comment:
this change doesn't make any sense to me. The `StorageType` for `UUIDType` is `&arrow.FixedSizeBinaryType{ByteWidth: 16}`. You wouldn't want the underlying storage type to be a different extension type. What's the use case where this would be correct over just using `!arrow.TypeEqual`?

Reply:
As I mentioned in the comment about the StorageEqual change, `storageType` passed in here is apparently an `ExtensionType`, so `TypeEqual()` doesn't work (as `FixedSizeBinaryType` isn't an extension type... which can be remedied by creating an extension on top of that type). We could also unrolling the `storageEqual()` into here as this is the only call site for that func, but I chose to leave it separate, thinking it could be useful for other test types in the future.